File: src/solutions/java/ValidParentheses.java
package solutions.java;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;
import java.util.Stack;

/*
# Problem
	- `Link`: https://leetcode.com/problems/valid-parentheses/
# Solution
	- `Author`: Kyungtaek Lim (Jonas)
	- `Date`: Mar 25, 2025
	- `Answer`: isValid / isValidAdvanced
*/
public class ValidParentheses {
	public static void main(String[] args) {
		System.out.println(isValid("()")); // true
		System.out.println(isValid("()[]{}")); // true
		System.out.println(isValid("(]")); // false
		System.out.println(isValid("([])")); // true
		
		System.out.println("---");
		
		System.out.println(isValidAdvanced("()")); // true
		System.out.println(isValidAdvanced("()[]{}")); // true
		System.out.println(isValidAdvanced("(]")); // false
		System.out.println(isValidAdvanced("([])")); // true
	}
	
	/*
    # Option #1
    - O(N)
	 */
    public static boolean isValid(String s) {
    	Deque<Character> deque = new ArrayDeque<>();
    	Character polledChar = null;
        for (char c: s.toCharArray()) {
        	if (c == ')') {
        		polledChar = deque.pollLast();
        		if (polledChar == null || '(' != polledChar) return false;
        	} else if (c == '}') {
        		polledChar = deque.pollLast();
        		if (polledChar == null || '{' != polledChar) return false;
        	} else if (c == ']') {
        		polledChar = deque.pollLast();
        		if (polledChar == null || '[' != polledChar) return false;
        	} else {
        		deque.add(c);
        	}
        }
        if (deque.isEmpty()) return true;
        return false;
    }
    
    /*
    # Option #2
    - O(N)
    - A little bit more organized source code than Option #1
     */
    public static boolean isValidAdvanced(String s) {
    	Stack<Character> stack = new Stack<>();
    	Map<Character, Character> map = Map.of(
    				')', '(',
    				']', '[',
    				'}', '{'
    			);
    	for (char c: s.toCharArray()) {
    		if (map.containsKey(c)) {
    			char top = stack.isEmpty() ? '#' : stack.pop();
    			if (top != map.get(c)) return false;
    		} else {
    			stack.push(c);
    		}
    	}
    	
    	return stack.isEmpty();
    }
}

File: src/solutions/python/ValidParentheses.py
'''
# Problem
	- `Link`: https://leetcode.com/problems/valid-parentheses/
# Solution
	- `Author`: Kyungtaek Lim (Jonas)
	- `Date`: Mar 25, 2025
	- `Answer`: isValid / isValidAdvanced
'''

class Solution:

    '''
    # Option #1
    - O(N)
    '''
    def isValid(self, s: str) -> bool:
        my_list = list()
        for c in s:
            if c == ')':
                if not my_list or my_list.pop() != '(':
                    return False
            elif c == '}':
                if not my_list or my_list.pop() != '{':
                    return False
            elif c == ']':
                if not my_list or my_list.pop() != '[':
                    return False
            else:
                my_list.append(c)
        return not my_list

    '''
    # Option #2
    - O(N)
    - A little bit more organized source code than Option #1
    '''
    def isValidAdvanced(self, s: str) -> bool:
        my_list = []
        my_dict = {'}': '{', ')': '(', ']': '['}
        
        for c in s:
            if c in my_dict:
                top = my_list.pop() if my_list else '#'
                if my_dict[c] != top:
                    return False
            else:
                my_list.append(c)
        
        return not my_list
            

if __name__ == '__main__':
    sol = Solution()
    print(sol.isValid("()")) # true
    print(sol.isValid("()[]{}")) # true
    print(sol.isValid("(]")) # false
    print(sol.isValid("([])")) # true

    print("---")

    print(sol.isValidAdvanced("()")) # true
    print(sol.isValidAdvanced("()[]{}")) # true
    print(sol.isValidAdvanced("(]")) # false
    print(sol.isValidAdvanced("([])")) # true